creator-node/src/ipfsClient.js

```diff
@@ -27,69 +25,121 @@ async function logIpfsPeerIds () {
-  let ipfsDaemonHashes
-  if (ENABLE_ASYNC_IPFS_ADD && !addToIPFSDaemon) {
-    ipfs.add(inputData, ipfsConfig) // Do not await it
+  // Generate hash with ipfs content hashing logic
+  const onlyHash = await ipfsHashOf(inputData)
+
+  // If async ipfs add is enabled, synchronously add to ipfs.
+  let ipfsDaemonHash
+  if (enableIPFSAdd) {
+    try {
+      ipfsDaemonHash = (await ipfs.add(inputData, ipfsConfig))[0].hash
+    } catch (e) {
+      logger.warn(`[ipfsClient - ipfsSingleAddWrapper()] Could not add content to ipfs. Defaulting to onlyHash=${onlyHash}: ${e.toString()}`)
+      return onlyHash
+    }
+  }
+
+  // Return the `ipfsDaemonHashes`, or `onlyHashes`. Prioritize `ipfsDaemonHashes`.
+  if (ipfsDaemonHash) {
+    logger.info(`[ipfsClient - ipfsSingleAddWrapper()] onlyHash=${onlyHash} ipfsDaemonHash=${ipfsDaemonHash} isSameHash=${onlyHash === ipfsDaemonHash}`)
+    return ipfsDaemonHash
   } else {
-    ipfsDaemonHashes = await ipfs.add(inputData, ipfsConfig)
+    logger.info(`[ipfsClient - ipfsSingleAddWrapper()] onlyHash=${onlyHash}`)
+    return onlyHash
   }
+}
+
+// TODO: Replace for images and figure out correct response structure
+/**
+ * Wrapper to ipfs.add() -- Allows enabling/disabling adding content to the ipfs daemon. Generally used for images (to generate dirs too)
+ * @param {Object} ipfs ipfs library -- can be two different versions (see ipfsClient.js)
+ * @param {Object[]} inputData an Object[] with the structure { path: string, content: buffer }
+ * @param {Object?} ipfsConfig ipfs add config options
+ * @param {Object?} logContext
+ * @param {boolean?} enableIPFSAdd flag to add content to ipfs daemon
+ * @returns {string|string[]|Object|Object[]} hashes from content addressing fn, or ipfs daemon responses
+ */
+// eslint-disable-next-line no-unused-vars
+async function ipfsMultipleAddWrapper (ipfs, inputData, ipfsConfig = {}, logContext = {}, enableIPFSAdd = false) {
+  const logger = genericLogger.child(logContext)
 
-  // If `buffers` is an array, get the content hash of each element. Else, it will be a single element.
-  // Also, if `buffers` is a single element, get the single element hash from `ipfsDaemonHashes`
+  // Generate hashes with ipfs content hashing logic
   let onlyHashes = []
-  if (Array.isArray(inputData)) {
-    // Note: See ipfs add code in `resizeImage.js` for `buffers` structure
-    for (const { content } of inputData) {
-      const hash = await ipfsHashOf(content)
-      onlyHashes.push(hash)
+  for (const { content } of inputData) {
+    const hash = await ipfsHashOf(content)
+    onlyHashes.push(hash)
+  }
+  const dirHash = await ipfsHashOf(inputData, {}, true)
+  onlyHashes.push(dirHash)
+
+  let ipfsDaemonResp
+  if (enableIPFSAdd) {
+    try {
+      ipfsDaemonResp = await ipfs.add(inputData, ipfsConfig)
+    } catch (e) {
+      logger.warn(`[ipfsClient - ipfsMultipleAddWrapper()] Could not add content to ipfs. Defaulting to onlyHash=${onlyHashes}: ${e.toString()}`)
```

Comment:
I wanted to discuss this case more -- before, when we added to ipfs and it threw an error, this method would fail. now, if adding to ipfs fails, we should default to the content hashing logic only and pass. this might case some issues when fetching metadata from discovery, but in a followup, we can also remove this dependency cc @hareeshnagaraj

Reply:
i'm open to this since discprov metadata retrieval will fallback to gateways, but want others feedback as well

might be worth tracking this very proactively, maybe through some counter field in healthcheck route that is tied to pingdom or some other approach. concerned if its just in logs we will forget about it

Reply:
i think we should keep this + track with some third party service cc @dmanjunath @raymondjacobson

Reply:
shouldn't discprov metadata retrieval be preferring content node gateways?

Reply:
I think the metadata retrieval should be preferring the CN gateways. but the code rn is only fetching from ipfs https://github.com/AudiusProject/audius-protocol/blob/master/discovery-provider/src/utils/ipfs_lib.py#L37

Reply:
actually, I think it is ok to not rethrow in the try/catch and actually return the content addressed hash bc the fallback is on the gateway https://github.com/AudiusProject/audius-protocol/blob/master/discovery-provider/src/utils/ipfs_lib.py#L37

Reply:
if we're stripping out IPFS so much here, why not fetch from content nodes first?